scenarios/validation_receipts/src/main.rs

```diff
@@ -61,82 +58,81 @@ fn agent_setup(
-            .pending_actions
-            .keys()
-            .cloned()
-            .collect::<Vec<_>>();
-
-        for action_hash in action_hash_list {
-            // for each action, get the validation receipts
-            let response: Vec<ValidationReceiptSet> = call_zome(
-                ctx,
-                "crud",
-                "get_sample_entry_validation_receipts",
-                action_hash.clone(),
-            )?;
-
-            for set in response.iter() {
-                let cur = *ctx
-                    .get_mut()
-                    .scenario_values
-                    .mut_op_complete(&action_hash, set.op_type.clone());
-
-                if set.receipts_complete && !cur {
-                    // if the action wasn't already complete report the time
-                    // and mark it complete
-                    reporter.add_custom(
-                        ReportMetric::new("validation_receipts_complete_time")
-                            .with_tag("op_type", set.op_type.clone())
-                            .with_tag("agent", agent.clone())
-                            .with_field("value", start.elapsed().as_secs_f64()),
-                    );
-                    *ctx.get_mut()
-                        .scenario_values
-                        .mut_op_complete(&action_hash, set.op_type.clone()) = true;
-                }
+    wait_for_receipts_for_action(ctx, &action_hash, start, wait_for_all)
+}
+
+/// Wait for validation receipts for a specific action.
+fn wait_for_receipts_for_action(
+    ctx: &mut AgentContext<HolochainRunnerContext, HolochainAgentContext<ScenarioValues>>,
+    action_hash: &ActionHash,
+    start: Instant,
+    wait_for_all: bool,
+) -> WindTunnelResult<()> {
+    // for each action, get the validation receipts
+    while !ctx.get().scenario_values.is_action_complete(action_hash) {
+        let response: Vec<ValidationReceiptSet> = call_zome(
+            ctx,
+            "crud",
+            "get_sample_entry_validation_receipts",
+            action_hash.clone(),
+        )?;
+        let reporter = ctx.runner_context().reporter();
+        let agent = ctx.get().cell_id().agent_pubkey().clone().to_string();
+
+        for set in response.iter() {
+            let cur = ctx
+                .get_mut()
+                .scenario_values
+                .mut_op_complete(action_hash, set.op_type.clone());
+
+            if set.receipts_complete && !*cur {
+                // if the action wasn't already complete report the time
+                // and mark it complete
+                reporter.add_custom(
+                    ReportMetric::new("validation_receipts_complete_time")
+                        .with_tag("op_type", set.op_type.clone())
+                        .with_tag("agent", agent.clone())
+                        .with_field("value", start.elapsed().as_secs_f64()),
+                );
+                *cur = true;
             }
         }
 
```

Comment:
Do we need to remove the action from `pending_actions` once we know that all the required receipts have been received?

Also do we want to track when all receipts for all op types have been received as another metric?

Also heads-up that I believe that currently the `Vec<ValidationReceiptSet>` only contains elements when we have received at least 1 ValidationReceipt for that op type. So we may also need to check here that the list length is equal to the number of op types created.